[export] Capture tensor.to() under export.

zhxchen17 opened the pull request:
Summary: We use to skip tensor.to() during tracing when the device is the same. This will bring some performance improvement in eager but making graph capture losing the semantics from original model. In this diff, we add an additional condition to skip the fast path when we don't have actual data inside a tensor, which is the case when we're using FakeTensor / FunctionalTensor to trace the model. This won't have perf impact on previous eager models while making sure we can capture the _to_copy() node in the graph.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r device_to

Differential Revision: D55969674

angelayi reviewed (approved):
ooh very cool

zhxchen17 commented:
test failures look fine to fix. will update this later

SherlockNoMad commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -388,7 +388,7 @@ static inline Tensor to_impl(
 
   // fast path
-  if (to_will_alias(self, dtype, layout, device, copy, optional_memory_format)) {
+  if (to_will_alias(self, dtype, layout, device, copy, optional_memory_format) && self.const_data_ptr()) {
```
Add a comment for const_data_ptr(), it's not trivial this is for fakeTensor/functionalTensor.

SherlockNoMad replied:
Also this is a proxy check for under FakeMode? 
maybe we can turn this into an explicit check for under FakeMode?

zou3519 reviewed (changes requested):
This changes the semantics of user code to make Tensor.to always copy under torch.export (and also torch.compile, it looks like?). I don't think that's what we want. A test case is the following (replace torch.compile with torch.export too)
```py
@torch.compile
def f(x):
   y = x.clone()
   z = y.to("cpu")
   y.sin_()
   return z

x = torch.randn(3)
out = f(x)
assert torch.allclose(out, x.sin())
```

zhxchen17 commented:
> This changes the semantics of user code to make Tensor.to always copy under torch.export (and also torch.compile, it looks like?). I don't think that's what we want. A test case is the following (replace torch.compile with torch.export too)
> 
> ```python
> @torch.compile
> def f(x):
>    y = x.clone()
>    z = y.to("cpu")
>    y.sin_()
>    return z
> 
> x = torch.randn(3)
> out = f(x)
> assert torch.allclose(out, x.sin())
> ```

@zou3519 I understand that in this particular case, we want to preserve the old behavior, so I don't think we necessarily want to dispatch to _to_copy always. Instead, we just want to preserve the device conversion semantics. I saw there's actually an aten::to which actually does this. https://fburl.com/code/8ax41xzx Could we redispatch to these ops instead?

zou3519 commented:
Is your eventual goal that you'd to export a graph that is device agnostic and the Tensor.to is one of the things that is preventing this from working?

zhxchen17 commented:
> Is your eventual goal that you'd to export a graph that is device agnostic and the Tensor.to is one of the things that is preventing this from working?

Yes.

zou3519 commented:
Our constraints are that we need to preserve the semantics of the user program. This is made complicated by how PT2 IR is functional -- a call to something like aten::to isn't functional and therefore it needs to be decomposed.

One middle ground is, when `out = tensor.to(...)` is called and is a no-op:
- IF we detect that it is safe to turn it into a copy, then we can emit a call to `aten::_to_copy`. It is safe to turn into a copy if `out` isn't being mutated or returned from the program and `tensor` isn't mutated. I'm not sure if this is practical enough for your use cases.

zhxchen17 commented:
> Our constraints are that we need to preserve the semantics of the user program. This is made complicated by how PT2 IR is functional -- a call to something like aten::to isn't functional and therefore it needs to be decomposed.
> 
> One middle ground is, when `out = tensor.to(...)` is called and is a no-op:
> 
> * IF we detect that it is safe to turn it into a copy, then we can emit a call to `aten::_to_copy`. It is safe to turn into a copy if `out` isn't being mutated or returned from the program and `tensor` isn't mutated. I'm not sure if this is practical enough for your use cases.

@zou3519 If we could do the middle ground approach you proposed, it will be helpful as well. I don't understand why capture the graph as 
```
y = aten::to.dtype_layout(tensor, device) # instead of _to_copy
```
will break the constraint of preserving user semantics with the presence of functionalization tho, especially in pre dispatch mode. It seems we can still preserve the aliasing relationship unlike _to_copy here, if we only reason about the pre condition and post condition of this call.

tugsbayasgalan commented:
@bdhirsh, @zhxchen17, and I talked offline. In general, it seems wrong behaviour to support mutation on the result of tensor.to because depending on whether you used CUDA tensor or CPU tensor, the behaviour of export program will be different. Consider following example:
```
y = x.to("cpu")
y.add_(5)
return x.sin()
```
In above code, we will mutate input if x is a CPU tensor while we won't if it is a CUDA tensor. As a result, i think it makes sense to just always decompose aten.to to aten._to_copy and ban the mutation on the output of aten.to.

zou3519 commented:
Banning the mutation seems good as long as we can raise an error on it

zhxchen17 commented:
@pytorchbot rebase

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -1902,6 +1902,48 @@ def _patch_config(kwargs):
+            self.assertIn(op, (torch.ops.aten._to_copy.default,))
+
+    def test_device_to_mutation(self):
+        class Module(torch.nn.Module):
```
I think we should fail on cases like this too:
```
def f(x):
    x1 = x.to(..)
    x2 = x1.to(..)
    return x2
 ```
 Maybe ok for now.

zhxchen17 commented:
@pytorchbot rebase

zhxchen17 committed:
[export] Capture tensor.to() under export. (#123732)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/123732

We use to skip tensor.to() during tracing when the device is the same. This will bring some performance improvement in eager but making graph capture losing the semantics from original model. In this diff, we add an additional condition to skip the fast path when we don't have actual data inside a tensor, which is the case when we're using FakeTensor / FunctionalTensor to trace the model. This won't have perf impact on previous eager models while making sure we can capture the _to_copy() node in the graph.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r device_to

Reviewed By: tugsbayasgalan, angelayi

Differential Revision: D55969674

zhxchen17 commented:
@pytorchbot merge
